### PR TITLE
fix: consider tuple type alone when checking type

### PR DESCRIPTION
**Pull request by xudong963**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

- Closes https://github.com/datafuselabs/databend/issues/12791

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/12814)

### CODE REVIEW

File: src/query/expression/src/block.rs [hunk 1]
```diff
@@ -590,3 +582,26 @@ impl Clone for Box<dyn BlockMetaInfo> {
...
 }
+
+fn check_type(data_type: &DataType, value: &Value<AnyType>) {
+    match value {
+        Value::Scalar(Scalar::Null) => {
+            assert!(data_type.is_nullable_or_null());
+        }
+        Value::Scalar(s) => {
+            match s {
+                Scalar::Tuple(ss) => {
```
Comment (b41sh) [inline src/query/expression/src/block.rs:593]: other nested types, such as `Array` and `Map` may have the same problem

Comment (xudong963) [reply src/query/expression/src/block.rs:593]: `Array` and `Map` nest `Column`, maybe don't have the problem.

Comment (b41sh) [reply src/query/expression/src/block.rs:593]: ok, `Array` and `Map` are ok
```sql
mysql> SELECT [true, tt1.b] FROM tt1 CROSS JOIN tt2;
+---------------+
| [true, tt1.b] |
+---------------+
| [1,1]         |
| [1,1]         |
| [1,1]         |
| [1,NULL]      |
| [1,NULL]      |
| [1,NULL]      |
| [1,0]         |
| [1,0]         |
| [1,0]         |
+---------------+
9 rows in set (0.04 sec)
```

**Review by b41sh — APPROVED** [review]

**Review by sundy-li — APPROVED** [review]

File: src/query/expression/src/block.rs [hunk 2]
```diff
@@ -590,3 +582,27 @@ impl Clone for Box<dyn BlockMetaInfo> {
...
+                    assert!(matches!(data_type, DataType::Tuple(_)));
+                    if let DataType::Tuple(dts) = data_type {
+                        for (s, dt) in ss.iter().zip(dts.iter()) {
+                            check_type(dt, &Value::Scalar(s.clone()));
+                        }
+                    }
+                }
+                _ => assert_eq!(s.as_ref().infer_data_type(), data_type.remove_nullable()),
+            }
+        }
```
Comment (andylokandy) [inline src/query/expression/src/block.rs:605]:
```suggestion
        Value::Scalar(Scalar::Tuple(fields)) => {
            // Check if data_type is Tuple type.
            let data_type = data_type.remove_nullable();
            assert!(matches!(data_type, DataType::Tuple(_)));
            if let DataType::Tuple(dts) = data_type {
                for (s, dt) in fields.iter().zip(dts.iter()) {
                    check_type(dt, &Value::Scalar(s.clone()));
                }
            }
        }
        Value::Scalar(s) => assert_eq!(s.as_ref().infer_data_type(), data_type.remove_nullable()),
```

**Pull request merged**